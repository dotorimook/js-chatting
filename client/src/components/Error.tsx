import React, { FC } from 'react';
import IComponentProps from 'interfaces/IComponentProps';
import ErrorModel from 'store/models/Error';
import styled from 'styled-components';
import Frame, { FrameCenter } from './Frame';
import img0 from 'assets/img/error/0.svg';
import img1 from 'assets/img/error/1.svg';
import Button from './form/Button';
import theme from 'style/theme';
import { inject, observer } from 'mobx-react';
import { applyFlexbox } from 'style/mixins';

interface IErrorProps extends IComponentProps{
  error?:ErrorModel;
};

const ErrorFrame = styled(FrameCenter)`
  position:fixed;
  background:none;
  transition: visibility opacity .1s ease;
  visibility: hidden;
  ${applyFlexbox('column','center','center')};
  z-index:99999;
  opacity: 0;
  &.visible {
    visibility: visible;
    opacity:1;
  }
  .msg,.title-img {
    z-index: 99999;
  }
  .msg {
    p {
      margin-bottom:2em;
    }
  }
`;
const Dimmer = styled(Frame)`
  position:fixed;
  top:0;
  left:0;
  background:${theme.dimmer};
`;

const ErrorComponent:FC<IErrorProps> = (props:IErrorProps) => {
  const {error} = props;
  if(!error)
    throw new Error('error model not found');
  return (
    <ErrorFrame className={`msg-frame ${error.visible? 'visible': ''}`}>
      <Dimmer/>
      <img className='title-img' src={[img0,img1][Math.round(Math.random()*1)]} alt='logo'/>
      <div className='msg'>
        <p>
          {error.msg}
        </p>
        <Button
          className='btn-close'
          onClick={()=>{error.visible = false;}}
        >
          알겠습니다.
        </Button>
      </div>
    </ErrorFrame>
  )
}

export default inject('error')(observer(ErrorComponent));