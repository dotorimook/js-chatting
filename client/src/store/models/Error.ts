import { observable, action } from "mobx";

interface IError {
  visible?: boolean;
  code?: string;
  msg?: string;
}

export default class Error {
  @observable visible?: boolean;
  @observable code?: string;
  @observable msg?: string;

  constructor(
    {visible = false, code='UNDEFINED', msg='알 수 없는 에러'}:IError
  ) {
    this.visible = visible;
    this.code = code;
    this.msg = msg;
  }

  @action
  toggle = () => {
    this.visible = !this.visible;
  }

  @action
  setError({visible = true, code, msg}:IError) {
    this.visible = visible;
    this.code = code;
    this.msg = msg;
  }
}