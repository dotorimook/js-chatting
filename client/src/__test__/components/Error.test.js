import React from 'react';
import 'setupTest';
import Error from 'components/Error';
import ErrorModel from 'store/models/Error';
import {mount} from 'enzyme';
import { Provider } from 'mobx-react';

describe('Error Test', () => {
  let ui = null;
  const error = new ErrorModel({});
  global.Math.random = jest.fn(()=>0);

  beforeAll(()=>{
    error.msg = 'error message';
    error.visible = true;
    ui = mount(
      <Provider error={error}>
        <Error/>
      </Provider>
    );
  });

  it('matches snapshot', () => {
    expect(ui).toMatchSnapshot();
  });

  it('properly renders error message', () => {
    expect(ui.find('.msg').text().indexOf(error.msg)).toBeGreaterThanOrEqual(0);
  });

  it('has a close button', () => {
    expect(ui.find('button.btn-close')).toHaveLength(1);
  });

  it('frame has visible className when error.visible true', () => {
    error.visible = true;
    ui.update();
    expect(ui.find('.msg-frame.visible').length).toBeGreaterThan(0);
  });

  it('error.visible becomes false called when a close button clicked', () => {
    const closeButton = ui.find('button.btn-close');
    closeButton.simulate('click');
    expect(error.visible).toBe(false);
  });
});