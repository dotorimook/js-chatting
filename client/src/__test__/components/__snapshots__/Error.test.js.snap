// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error Test matches snapshot 1`] = `
<MobXProvider
  error={
    Error {
      "code": "UNDEFINED",
      "msg": "error message",
      "visible": true,
    }
  }
>
  <inject-with-error(ErrorComponent)>
    <ErrorComponent
      error={
        Error {
          "code": "UNDEFINED",
          "msg": "error message",
          "visible": true,
        }
      }
    >
      <Styled(Styled(styled.div))
        className="msg-frame visible"
      >
        <StyledComponent
          className="msg-frame visible"
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [],
              "componentStyle": ComponentStyle {
                "componentId": "sc-EHOje",
                "isStatic": true,
                "lastClassName": "dijeET",
                "rules": Array [
                  "
  width:100%;
  height:100%;
  ",
                  "
  
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;

  
  -webkit-flex-wrap: nowrap;
      -moz-flex-wrap: nowrap;
      -ms-flex-wrap: nowrap;
          flex-wrap: nowrap;

  
  -webkit-flex-direction: column;
      -moz-flex-direction: column;
      -ms-flex-direction: column;
          flex-direction: column;

  
  -webkit-justify-content: flex-start;
      -moz-justify-content: flex-start;
      -ms-justify-content: flex-start;
          justify-content: flex-start;
            -ms-flex-pack: flex-start;

  
  -webkit-align-items: stretch;
      -moz-align-items: stretch;
      -ms-align-items: stretch;
          align-items: stretch;

",
                  ";
  
  &.msg-frame {
    .title-img {
      width: 50%;
      display:block;
      margin: 0 auto;
    }
    .msg {
      margin-top:2em;
      text-align:center;
    }
  }
",
                  "
  ",
                  "
  
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;

  
  -webkit-flex-wrap: nowrap;
      -moz-flex-wrap: nowrap;
      -ms-flex-wrap: nowrap;
          flex-wrap: nowrap;

  
  -webkit-flex-direction: column;
      -moz-flex-direction: column;
      -ms-flex-direction: column;
          flex-direction: column;

  
  -webkit-justify-content: center;
      -moz-justify-content: center;
      -ms-justify-content: center;
          justify-content: center;
            -ms-flex-pack: center;

  
  -webkit-align-items: stretch;
      -moz-align-items: stretch;
      -ms-align-items: stretch;
          align-items: stretch;

",
                  ";
",
                  "
  position:fixed;
  background:none;
  transition: visibility opacity .1s ease;
  visibility: hidden;
  ",
                  "
  
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;

  
  -webkit-flex-wrap: nowrap;
      -moz-flex-wrap: nowrap;
      -ms-flex-wrap: nowrap;
          flex-wrap: nowrap;

  
  -webkit-flex-direction: column;
      -moz-flex-direction: column;
      -ms-flex-direction: column;
          flex-direction: column;

  
  -webkit-justify-content: center;
      -moz-justify-content: center;
      -ms-justify-content: center;
          justify-content: center;
            -ms-flex-pack: center;

  
  -webkit-align-items: center;
      -moz-align-items: center;
      -ms-align-items: center;
          align-items: center;

",
                  ";
  z-index:99999;
  opacity: 0;
  &.visible {
    visibility: visible;
    opacity:1;
  }
  .msg,.title-img {
    z-index: 99999;
  }
  .msg {
    p {
      margin-bottom:2em;
    }
  }
",
                ],
              },
              "displayName": "Styled(Styled(styled.div))",
              "foldedComponentIds": Array [
                "sc-bdVaJa",
                "sc-bwzfXH",
              ],
              "render": [Function],
              "styledComponentId": "sc-EHOje",
              "target": "div",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <div
            className="sc-bdVaJa sc-bwzfXH msg-frame visible sc-EHOje dijeET"
          >
            <Styled(styled.div)>
              <StyledComponent
                forwardedComponent={
                  Object {
                    "$$typeof": Symbol(react.forward_ref),
                    "attrs": Array [],
                    "componentStyle": ComponentStyle {
                      "componentId": "sc-bZQynM",
                      "isStatic": true,
                      "lastClassName": "cNPsHz",
                      "rules": Array [
                        "
  width:100%;
  height:100%;
  ",
                        "
  
  display: -webkit-box;
  display: -moz-box;
  display: -ms-flexbox;
  display: -webkit-flex;
  display: flex;

  
  -webkit-flex-wrap: nowrap;
      -moz-flex-wrap: nowrap;
      -ms-flex-wrap: nowrap;
          flex-wrap: nowrap;

  
  -webkit-flex-direction: column;
      -moz-flex-direction: column;
      -ms-flex-direction: column;
          flex-direction: column;

  
  -webkit-justify-content: flex-start;
      -moz-justify-content: flex-start;
      -ms-justify-content: flex-start;
          justify-content: flex-start;
            -ms-flex-pack: flex-start;

  
  -webkit-align-items: stretch;
      -moz-align-items: stretch;
      -ms-align-items: stretch;
          align-items: stretch;

",
                        ";
  
  &.msg-frame {
    .title-img {
      width: 50%;
      display:block;
      margin: 0 auto;
    }
    .msg {
      margin-top:2em;
      text-align:center;
    }
  }
",
                        "
  position:fixed;
  top:0;
  left:0;
  background:",
                        "rgba(255,255,255,0.95)",
                        ";
",
                      ],
                    },
                    "displayName": "Styled(styled.div)",
                    "foldedComponentIds": Array [
                      "sc-bdVaJa",
                    ],
                    "render": [Function],
                    "styledComponentId": "sc-bZQynM",
                    "target": "div",
                    "toString": [Function],
                    "warnTooManyClasses": [Function],
                    "withComponent": [Function],
                  }
                }
                forwardedRef={null}
              >
                <div
                  className="sc-bdVaJa sc-bZQynM cNPsHz"
                />
              </StyledComponent>
            </Styled(styled.div)>
            <img
              alt="logo"
              className="title-img"
              src="0.svg"
            />
            <div
              className="msg"
            >
              <p>
                error message
              </p>
              <Button
                className="btn-close"
                onClick={[Function]}
              >
                <styled.button
                  className="btn-close"
                  onClick={[Function]}
                >
                  <StyledComponent
                    className="btn-close"
                    forwardedComponent={
                      Object {
                        "$$typeof": Symbol(react.forward_ref),
                        "attrs": Array [],
                        "componentStyle": ComponentStyle {
                          "componentId": "sc-ifAKCX",
                          "isStatic": true,
                          "lastClassName": "bZsaUo",
                          "rules": Array [
                            "
  width: 100%;
  font-size:1em;
  border:none;
  padding: 1em 1em;
  background-color:transparent;

  .material-icons {
    vertical-align:bottom;
    &.left {
      margin-right:.2em;
    }
  }
  
  transition: all 0.2s ease;
  &:hover {
    background-color:",
                            "rgba(184, 242, 230, 1)",
                            "
  }
  &:active {
    transform:translateY(2px);
  }
  
  &[type=submit] {
    background-color:",
                            "rgba(184, 242, 230, 1)",
                            ";
    &:hover {
      background-color:",
                            "rgba(250, 243, 221, 1)",
                            ";
    }
  }
",
                          ],
                        },
                        "displayName": "styled.button",
                        "foldedComponentIds": Array [],
                        "render": [Function],
                        "styledComponentId": "sc-ifAKCX",
                        "target": "button",
                        "toString": [Function],
                        "warnTooManyClasses": [Function],
                        "withComponent": [Function],
                      }
                    }
                    forwardedRef={null}
                    onClick={[Function]}
                  >
                    <button
                      className="btn-close sc-ifAKCX bZsaUo"
                      onClick={[Function]}
                    >
                      알겠습니다.
                    </button>
                  </StyledComponent>
                </styled.button>
              </Button>
            </div>
          </div>
        </StyledComponent>
      </Styled(Styled(styled.div))>
    </ErrorComponent>
  </inject-with-error(ErrorComponent)>
</MobXProvider>
`;
